lib/base64_attach.rb

```diff
@@ -0,0 +1,13 @@
+# Helper method to decode a base64 file and create a StringIO file
+module Base64Attach
+  module_function
+
+  def attachment_from_base64(data:, filename: Time.current.to_i.to_s)
+    base64_image = data.sub(/^data:.*,/, '')
+
+    decoded_image = Base64.decode64(base64_image)
+    image_io = StringIO.new(decoded_image)
+
+    { io: image_io, filename: filename }
+  end
+end
```

Comment:
I have found this method not to work when doing something like:
`user.avatar.attach(attachment_from_base64(params[:user][:avatar])` which I think is the expected way to use this.

I found by testing on my Target app that there's some more processing to do before being able to attach a base64 image like that. I just get an error from ActiveStorage doing so.

This is a code I came up with after a lot trial and error and research that does work for me:
```ruby
def attachment_from_data(data)
  split_base64 = splitBase64(data)
  extension = split_base64[:extension]
  data2 = split_base64[:data]
  io = io_from_data(data2)
  file = create_tempfile(io, extension)
  { io: File.open(file.path), filename: '' }
end

def create_tempfile(io, extension)
  file = Tempfile.new(["from_base_64", extension])
  file.binmode
  file.write io.read
  file.close
  file
end

def io_from_data(data)
  decoded_image = Base64.decode64(data)
  strIO = StringIO.new(decoded_image)
end

def splitBase64(uri)
  if uri.match(%r{^data:(.*?);(.*?),(.*)$})
    return {
      type:      $1, # "image/png"
      encoder:   $2, # "base64"
      data:      $3, # data string
      extension: $1.split('/')[1] # "png"
    }
  end
end
```

This way by doing `user.avatar.attach(attachment_from_data(params[:user][:avatar])` it does work.

What do others think?

Reply:
I think the reason of the failure is that the code you posted is not valid. It should be `user.avatar.attach(attachment_from_base64(data: params[:user][:avatar])` note the `data:` . Would you mind also posting the error you get?

Reply:
Nevermind, passing the data like that works.

Is that because of the definition of the method? If you define a method with the params names with a colon afterwards it means that you have to specify what you are passing to the method?

Reply:
Exactly, they are called keyword arguments https://robots.thoughtbot.com/ruby-2-keyword-arguments

Reply:
Great! Thanks 👍